packages/flutter/test/cupertino/switch_test.dart

```diff
@@ -80,6 +80,117 @@ void main() {
+                      setState(() {
+                        value = newValue;
+                      });
+                    },
+                  ),
+                  CupertinoSwitch(
+                    key: switchKey2,
+                    value: value2,
+                    onChanged: (bool newValue) {
+                      setState(() {
+                        value2 = newValue;
+                      });
+                    },
+                  ),
+                ]
+              ),
+            );
+          },
+        ),
+      ),
+    );
+
+    await tester.tap(find.byKey(switchKey));
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(1));
+    expect(log[0], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+
+    await tester.tap(find.byKey(switchKey2));
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(2));
+    expect(log[1], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+
+    await tester.tap(find.byKey(switchKey));
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(3));
+    expect(log[2], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+
+    await tester.tap(find.byKey(switchKey2));
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(4));
+    expect(log[3], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+    debugDefaultTargetPlatformOverride = null;
+  });
+
+  testWidgets('Haptic vibration triggers on drag', (WidgetTester tester) async {
+    debugDefaultTargetPlatformOverride = TargetPlatform.iOS;
+    bool value = false;
+    final List<MethodCall> log = <MethodCall>[];
+
+    SystemChannels.platform.setMockMethodCallHandler((MethodCall methodCall) async {
+      log.add(methodCall);
+    });
+
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: StatefulBuilder(
+          builder: (BuildContext context, StateSetter setState) {
+            return Center(
+              child: CupertinoSwitch(
+                value: value,
+                dragStartBehavior: DragStartBehavior.down,
+                onChanged: (bool newValue) {
+                  setState(() {
+                    value = newValue;
+                  });
+                },
+              ),
+            );
+          },
+        ),
+      ),
+    );
+
+    await tester.drag(find.byType(CupertinoSwitch), const Offset(30.0, 0.0));
+    expect(value, isTrue);
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(1));
+    expect(log[0], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+    debugDefaultTargetPlatformOverride = null;
+  });
+
   testWidgets('Switch can drag (LTR)', (WidgetTester tester) async {
```

Comment:
Can you also check that the knock doesn't happen when the parent widget programmatically changes the current value. Only on user interaction?